src/shared_utilities/utils.jl

```diff
@@ -666,6 +666,65 @@ function NaNCheckCallback(
     SciMLBase.DiscreteCallback(cond, affect!)
 end
 
+
+"""
+    FrequencyBasedCallback(
+        frequency::Union{AbstractFloat, Dates.Period},
+        start_date,
+        dt;
+        func,
+        func_args...
+    ) -> DiscreteCallback
+
+Constructs a `DiscreteCallback` that calls a given function with specified
+arguments at a specified frequency in simulation time.
+
+# Arguments
+- `frequency`: Either a `Float` (assumed to be in seconds) or a `Dates.Period`
+  (e.g., `Hour(6)`) indicating how often to trigger the callback.
+- `start_date`: The calendar start time of the simulation.
+- `dt`: The model timestep (used for divisibility warning).
+- `func`: A function that is called every time the schedule triggers.
+- `func_args...`: Additional arguments passed to `func`.
+
+The callback uses `ClimaDiagnostics.EveryCalendarDtSchedule` to determine when to
+call the function based on the `frequency`. 
+"""
+function FrequencyBasedCallback(
+    frequency::Union{AbstractFloat, Dates.Period},
+    start_date,
```

Comment:
It would be good to take in t0 (or t_start), since the simulation starts at `start_date + Seconds(t0)` if t0 != 0, like how we do for CheckpointCallback.

Also, could you open an issue to use this instead of CheckpointCallback? This is more general so we can use it for that case too (but it doesn't have to be done in this PR)